contract/src/offer-up.contract.js

```diff
@@ -1,70 +1,143 @@
-/** @file Contract to mint and sell Item NFTs. */
+/**
+ * @file Contract to mint and sell a few Item NFTs at a time.
+ *
+ * We declare variables (including functions) before using them,
+ * so you may want to skip ahead and come back to some details.
+ * @see {start} for the main contract entrypoint
+ *
+ * @see {@link https://docs.agoric.com/guides/zoe/|Zoe Overview} for a walk-thru of this contract
+ * @see {@link https://docs.agoric.com/guides/js-programming/hardened-js.html|Hardened JavaScript}
+ * for background on `harden` and `assert`.
+ */
 // @ts-check
 
 import { Far } from '@endo/far';
 import { M, getCopyBagEntries } from '@endo/patterns';
 import { AmountMath, AssetKind } from '@agoric/ertp/src/amountMath.js';
 import { AmountShape } from '@agoric/ertp/src/typeGuards.js';
-// Use the deprecated atomicRearrange API
-// for compatibility with mainnet1B.
+import { atomicRearrange } from '@agoric/zoe/src/contractSupport/atomicTransfer.js';
 import '@agoric/zoe/exported.js';
 
-import { makeTracer } from './debug.js';
-
 const { Fail, quote: q } = assert;
 
-const trace = makeTracer('OfferUp', true);
+// #region bag utilities
+/** @type { (xs: bigint[]) => bigint } */
+const sum = xs => xs.reduce((acc, x) => acc + x, 0n);
 
-/** @param {Amount<'copyBag'>} amt */
-const bagValueSize = amt => {
-  /** @type {[unknown, bigint][]} */
-  const entries = getCopyBagEntries(amt.value); // XXX getCopyBagEntries returns any???
-  const total = entries.reduce((acc, [_, qty]) => acc + qty, 0n);
-  return total;
+/**
+ * @param {import('@endo/patterns').CopyBag} bag
+ * @returns {bigint[]}
+ */
+const bagCounts = bag => {
+  const entries = getCopyBagEntries(bag);
+  return entries.map(([_k, ct]) => ct);
 };
+// #endregion
 
 /**
- * @param {ZCF<{tradePrice: Amount}>} zcf
+ * This contract is parameterized by terms for price and,
+ * optionally, a maximum number of items sold for that price (default: 3).
+ *
+ * @typedef {{
+ *   tradePrice: Amount;
+ *   maxItems?: bigint;
+ * }} ItemTerms
+ */
+
+export const meta = {
+  customTermsShape: M.splitRecord(
+    { tradePrice: AmountShape },
+    { maxItems: M.bigint() },
+  ),
+};
+// compatibility with an earlier contract metadata API
```

Comment:
What does this mean? I have not seen `meta.customTermsShape` before.

Reply:
It's like a proposal shape but for terms.

Documentation is on the TODO list:

 - https://github.com/Agoric/documentation/issues/743

That issue has pointers to context.